# 에러/예외 처리

> 1. 디버깅
> 2. 에러와 예외
> 3. 예외 처리
> 4. 예외 발생 시키기



## 1. 디버깅 🐞

- branches 

  모든 조건이 원하는대로 동작하는지

- for loops 

  반복문에 진입하는지, 원하는 횟수만큼 실행되는지

- while loops 

  for loops 와 동일, 종료조건이 제대로 동작하는지 

- function

  함수 호출시, 함수 파라미터, 함수 결과 



## 2. 에러와 예외 

- 문법 에러(Syntax Error)

  - syntax error 가 발생하면, 파이썬 프로그램은 실행이 되지 않음
  - 파일이름, 줄번호, ^ 문자를 통해 파이썬이 코드를 읽어 나갈 때 문제가 발생한 위치 표현
  - 줄에서 에러가 감지된 가장 앞의 위치를 가리키는 캐럿 기호(^) 를 표시 

- 예외(Exception)

  - 실행 도중 예상치 못한 상황을 맞이하면, 프로그램 실행을 멈춤

    • 문장이나 표현식이 문법적으로 올바르더라도 발생하는 에러

  - 실행 중에 감지되는 에러들을 예외(Exception)라고 부름

  - 예외는 여러 타입(type)으로 나타나고, 타입이 메시지의 일부로 출력됨 • NameError, TypeError 등은 발생한 예외 타입의 종류(이름)

  - 모든 내장 예외는 Exception Class를 상속받아 이뤄짐

  - 사용자정의예외를만들어관리할수있음		

  #### 예외의 종류

  - ZeroDivisionError : 0으로 나눌 경우

    ```python
    10/0
    
      File "/Users/mac/Desktop/TIL/KDT-python/0718_1.py", line 1, in <module>
        10/0
    ZeroDivisionError: division by zero
    ```

  - NameError : namespace 상에 이름이 없는 경우

    ```python
    print(name_error)
    
      File "/Users/mac/Desktop/TIL/KDT-python/0718_1.py", line 1, in <module>
        print(name_error)
    NameError: name 'name_error' is not defined
    ```

    

  #### 타입 불일치(TypeError)

  ```python
  # int와 str을 더함
  		1 + '1'
  TypeError: unsupported operand type(s) for +: 'int' and 'str'
  # str을 반올림    
      round('3.5')
  TypeError: type str doesn't define __round__ method
  # 0개의 인자를 넣는경우
      divmod()
  TypeError: divmod expected 2 arguments, got 0
  # 인자 개수 초과
      divmod(1, 2, 3)
  TypeError: divmod expected 2 arguments, got 3
  ```

  

  #### 타입은 올바르나 값이 적절하지 않거나 없는경우(ValueError)

  ```python
  		int('3.5')
  ValueError: invalid literal for int() with base 10: '3.5'
      
      range(3).index(6)
  ValueError: 6 is not in range
  ```

  

  #### indexError

  ```python
      empty_list[2]
  IndexError: list index out of range
  ```

  

  #### KeyError

  ```python
      song ['BTS']
  KeyError: 'BTS'
  ```

  

  #### 존재하지 않는 모듈을 import 하는 경우(ModuleNotFoundError)

  ```python
      import nonamed
  ModuleNotFoundError: No module named 'nonamed'
  ```

  

  #### 모듈은 있으나 존재하지 않는 클래스/함수를 가져오는 경우(importError)

  ```python
      from random import samp
  ImportError: cannot import name 'samp' from 'random' (/Users/mac/.pyenv/versions/3.9.13/lib/python3.9/random.py)
  ```

  

  #### 들여쓰기가 적절하지 않은 경우(IndentationError)

  ```python
      print(i)
      ^
  IndentationError: expected an indented block
  ```

  

  #### 임의로 프로그램을 종료하였을때(KeyboardInterrupt)

  ```python
  while True:
    continue
  # 임의로 종료함  
  KeyboardInterrupt :
  ```



## 3. 예외처리

- try 문(statement) / except 절(clause)을 이용하여 예외 처리를 할 수 있음

-  try문
  - 오류가발생할가능성이있는코드를실행
  - 예외가 발생되지 않으면, except 없이 실행 종료 
- Except 문 
  - 예외가 발생하면, except 절이 실행
  - 예외상황을처리하는코드를받아서적절한조치를취함

예외처리 예시

```python
# 숫자 입력을 받아서 출력
numbers = input('숫자를 입력해주세요:')

try:
  print(100/int(number))
except ZeroDivisionError as err: # 각각 예외처리를 할수있음
  print(err) # as err -> print(err) 파이썬에서 실제로 출력되는 에러를 출력할수 있음
  print('0으로 나눌 수는 없습니다.')
except ValueError:
  print('숫자 형식을 입력해주세요.')
except Exception: 
  print('오류')
finally: # 예외 여부와 관계없이 항상 실행
  print('입력을 종료합니다.')
```

#### 정리

- try
  코드를실행함

- except
   try문에서예외가발생시실행함

- else
  try 문에서 예외가 발생하지 않으면 실행함

- finally
  예외발생여부와관계없이항상실행함



## 4. 예외 발생시키기

- raise를 통해 예외를 강제로 발생

  raise <표현식>(메시지)

  ```python
   raise
  # -------
  # RuntimeError Traceback (most recent call last) # ----> 1 raise
  # RuntimeError: No active exception to reraise
  ```

  